### DISCUSSION TITLE
[Proposal] Add Customization System

### DISCUSSION THREAD

**Discussion by sdilts** [stumpwm/stumpwm#1112] [Ideas]
# Proposal: Add Customization System

This is a proposal to add a customization system to StumpWM to make it easier view customization options and ensure that the values that the customizable variables are set to are correct. Inspired by the discussion in #832.

## Rational
StumpWM has a large amount of configuration options that are available to users. You can change things like mouse behavior, colors, the mode line, window appearance, and others. While most of these customization can be accomplished by setting global variables, a configuration system to ensure that these features are self-documenting, self-validating, and easy to use would make it easier for users to  configure their system.

The configuration system would allow users to do three main things: examine available settings, view each setting's default values, and set the values to a valid value.

### Examine Customizable Settings
It should be possible to retrieve a list of all the customizable settings using a function. This would return a list of all of the available customizable variables along with their valid values / validator functions. This function might have optional parameters that allow you to filter the returned list.

It should also be possible to examine an individual setting to view it's default value,  current value, and documentation. An added bonus would also be the ability to get the information that is used to validate variable's value. Two functions should exist, one that gets the information in a data structure and one to get the information as a printable string. Edit: These functions should also return the setting's current value.

### Setting Customizations
There should be a way to set the customizable variables  so that the value is validated before the variable is changed. When the given value is incorrect, an error condition should be raised.

### Create new Customizable Settings
There should be at least one `defparameter` style macro that registers a setting, it's documentation, as well as validation information. Since it is probably a very common use case, it would be nice to have a version that accepts a list of possible values that the setting can hold.

### Grouping variables
While this feature isn't as important as the other ones and optional, it would be nice to be able to categorize settings to make them more searchable or separate them out on a per-module basis.

### Implemented in a separate package
This should be implemented in it's own package that doesn't depend on any of StumpWM's internals so that it is easier to reuse and not coupled to the current implementation.

**Comment by szos** [comment]
I've started work on a more robust system than the one talked about in #832 ([located here, check branch functionize](https://github.com/szos/defcustom/tree/functionize)) and I'm running into some design issues. I would like to make this system more general and allow for validation of not only dynamic places but of slots and the like as well, both through accessors and slot-value/with-slots. The previous system works fine for symbols (and symbols only) and can easily be translated into the system currently being written, but when it comes to extending it like I mentioned everything falls apart. I'd love to get some input from people more knowledgeable than myself on how this should be designed.  

do note that that system has been thrown together over the course of a day or two and has MANY examples of terrible coding.

**Comment by montaropdf** [comment]
Waw, I am really impressed by the work done in such a small amount of time, after I open #832.

I would like to add something to the reflection on the design. Have you considered storing the data into a database?

In this design, the database would be loaded before the user configuration. It would provide interactive functions/commands to modify its content instead of setting the variables through the configuration process. The hope is to reduce the global load time of StumpWM, in addition to adding constraint to the possible values of each variables and improving the structure of the user configuration.

I want to stress out that, I am speaking of a database as a concept and not as a particular implementation. The format of the database needs to be decided based on the needs of the project and the users.

The way I see the process of using the database is:
1. StumpWM start
2. The database module is loaded
3. The content of the database is exported to lisp code, if changes to the database happens after the last lisp code generation
4. The lisp code is evaluated
5. The user configuration is loaded

The end-user interface to the database could be:
- Exporting the content as lisp code (and possibly open the preferred text editor)
- Importing the lisp code to the DB with validation of the content
- Checking the DB integrity
- Getting the value of a variable
- Setting the value of a variable
- Searching for variables
- Displaying the documentation for a variable

Reading the various message exchanges in #839, it seems that the above idea is partly implemented, however, my intent is to modify and store the variables using interactive functions or commands, instead of scattering setters through the user configuration. The database doesn't prevent from modifying, manually, the variables, through the configuration, so, it should be backward compatible.

With regards to the interactive functions/commands to modify a variable, I don't know much how to present the data on the screen. One possibility, could be to open the preferred editor of the user, with the lisp form of the variable and its accompanying documentation as comments. This would be similar to things like the behavior of the command `git commit` without the `-m` flag.

**Comment by xsteve** [comment]
Hi Roland,

whatever is going to be implemented:

For me it is important that this database is not mandatory. I don't see
a benefit for my workflow.

I like to have a .stumpwmrc that configures all of stumpwm without any
state that is stored in a binary database.

I don't think that the startup time of stumpwm is a problem - it is not for me.
How many milliseconds do you think that you save that way? Why is this important?

Regards
  Stefan.

**Comment by sdilts** [comment]
@montaropdf If I understand what you want correctly, you'd like to have a non-code solution to configuring the settings. Instead of having to edit the rc file, re-execute it, then test it to see if you like what changed, you'd like to be able to set the individual values interactively using various commands, then save that configuration permanently? From the functionality listed here, it would be possible to build something that functions similarly to [emac's customization system](http://ergoemacs.org/emacs/emacs_custom_system.html) that generates code an inserts it into the user's rc file.

The one thing that is missing from the proposal is a way to view each setting's value via a function instead of examining the underlying variable itself. I'll add it in shortly. Almost everything else is covered: 

> Exporting the content as lisp code (and possibly open the preferred text editor)

Once you can examine a settings value via a function, you could use the stuff listed in the `Examine Customizable Settings` section to implement this.

> Importing the lisp code to the DB with validation of the content
   Checking the DB integrity

The functions that set the settings do this already, so nothing needs to be added (see `set-configuration-atomic` in #839). 

> Getting the value of a variable interactively
   Setting the value of a variable interactively
   Searching for variables interactively
   Displaying the documentation for a variable interactively

I added "interactively" to these, and I'm guessing that's what you meant. Once you can examine a settings value via a function, all of this is possible via the proposed API. Doing this directly in the configuration system code would break the requirement of not relying on anything in the Stumpwm package, as we would have to call stumpwm-specific functions to do this. I'm hesitant to break that requirement, but it would be easy to write some wrapper macros in the StumpWM package.

> The hope is to reduce the global load time of StumpWM

The impacts of this system to startup time could get costly due to running all of the validation functions at startup, but it isn't something I would worry about until it actually makes a noticeable difference. Once we get the basic system in place, then we can worry about startup time.

**Comment by szos** [comment]
@xsteve Hi Stefan, 

With regards to the systems currently being implemented (both the ´defcustom´ system and the ´config-system´ system), they are designed to function alongside setf and not to replace setf. As far as avoiding a database, The current thought-train has been to create database entries alongside dynamic variable definitions (see `define-config-parameter` and co. in the draft sdilts submitted), But i think it wouldnt be terribly difficult to build a method of suppressing this db generation (perhaps with a dynamic variable, or with a flag when compiling? something like ´make --no-db´?)

@montaropdf  Hi Roland,

I think it would be quite nice to load the db after StumpWM starts, but I'm unsure of how to do this without decoupling variable definitions and config definitions. One of the goals has been to create a macro that will create a variable or parameter alongside a config-info object (in keeping with the name scheme of the draft), as then definitions could be implemented incrementally and we wouldnt need two forms per variable/parameter definition. One option could be to have these definition forms not create a config-info object but instead write their definitions to a file which could then be loaded at the users convenience. this would also let people who dont want this system to avoid loading it altogether. I'm not terribly familiar with databases though, so perhaps ive misunderstood you.

as far as using interactive commands instead of setters, I think that is something that would/should be implemented in stumpwm propper, instead of in the customization package - I think the customization package should be as general as possible to encourage reuse in similar situations. 

When it comes to displaying information to the user, given that there has been some talk of integrating CLIM for use in the mode line and input bar, I'm tempted to say that this could be implemented as a CLIM frame, with commands that utilize the database for validation. I think such an application would be decently simple, and a minimal version would only need an interactor pane. A non-minimal might define a presentation type for database entries and use that to create a graphical interface for customizing things.

**Comment by montaropdf** [comment]
> @montaropdf If I understand what you want correctly, you'd like to have a non-code solution to configuring the settings. Instead of having to edit the rc file, re-execute it, then test it to see if you like what changed, you'd like to be able to set the individual values interactively using various commands, then save that configuration permanently?

This is exactly what I had in mind. And to come back to my idea of a Stump Desktop Environment (for which I still need to open an issue or at least publish a post on the ML), One could think about providing GUIs dedicated to sets of variables, but that's another topic entirely.

**Comment by montaropdf** [comment]
> @montaropdf Hi Roland,
> 
> I think it would be quite nice to load the db after StumpWM starts, but I'm unsure of how to do this without decoupling variable definitions and config definitions. One of the goals has been to create a macro that will create a variable or parameter alongside a config-info object (in keeping with the name scheme of the draft), as then definitions could be implemented incrementally and we wouldnt need two forms per variable/parameter definition. One option could be to have these definition forms not create a config-info object but instead write their definitions to a file which could then be loaded at the users convenience. this would also let people who dont want this system to avoid loading it altogether. I'm not terribly familiar with databases though, so perhaps ive misunderstood you.

Once again, currently, I am thinking of a database as a concept and not a particular implementation. This means that to me in its concrete form, actually a db could be a set of text or binary files in a directory or a big binary blob like a postgreSQL or mariaDB RDBMS or a NoSQL DB like MongoDB, to name a few. We need, first, to settle on the design, workflow, use cases, and maybe the UI, before thinking about the database implementation to use to store the data.

I hope this makes thinks clearer to everybody. I think it is too early to consider a particular implementation over another at this stage.
 
> as far as using interactive commands instead of setters, I think that is something that would/should be implemented in stumpwm propper, instead of in the customization package - I think the customization package should be as general as possible to encourage reuse in similar situations.

I am all for it.

> When it comes to displaying information to the user, given that there has been some talk of integrating CLIM for use in the mode line and input bar, I'm tempted to say that this could be implemented as a CLIM frame, with commands that utilize the database for validation. I think such an application would be decently simple, and a minimal version would only need an interactor pane. A non-minimal might define a presentation type for database entries and use that to create a graphical interface for customizing things.

That would be effectively great to capitalize on this initiative.

**Comment by fourier** [comment]
It sounds more like a complication. The rationale states that `The configuration system would allow users to do three main things: examine available settings, view each setting's default values, and set the values to a valid value.`. 
We can examine available settings already (listing all public variables in the package), check the value in SLIME/SLY, set it. The Only thing not available is a default value - which should be in a docstring for the variable theoretically.
Current system is simple and effective for any CL user (they are the target group for WM) and maintains the spirit of simplicity.

**Comment by PuercoPop** [comment]
> It sounds more like a complication.

@fourier Thanks for voicing your concerns and I agree that we have to move carefuly to avoid creating a complex system that gets in the way of people who just want to `setf` away and be done with it (:raising_hand:). It doesn't have to be much more than book-keeping over def{var,parameter}s, an *optional* setter that does some validation before setting the variable and an UI as a contrib. In the initial PR I'm arguing that we only replace the calls to defvar/defparameter with a defconfig macro that just expands to defvar/defparameter, not even doing book-keeping. 

> We can examine available settings already (listing all public variables in the package)

Not all public variables are settings.

> Current system is simple and effective for any CL user

And it will continue to work for those who don't want it. Just like any emacs user can ignore the `customize` system (and many do).

> check the value in SLIME/SLY, set it

Sure, we don't even need SLY, we can call eval directly from StumpWM.

**Comment by dmb2** [comment]
I've been passively keeping track of this discussion and I think I can summarize my opinion as:
- We can't break existing configs/setf(setq) should be good enough.  This is familiar to emacs users where customize has a UI but it ultimately writes a custom section to .emacs that is no different than writing it yourself.
- Any G/UI implementations should be done in a contrib module, and any core changes (macros etc) should be included for users to run in their rc files independent of the UI contrib module.
- I'm very supportive of this new direction in features/functionality.

**Comment by szos** [comment]
I've been thinking about this and eventually got around to implementing something. Its done in a way that would be a module, not implemented in StumpWM proper. 

Its broken up into two repos, one to implement the GUI, one to handle the database and more general functionality. These are located [here](https://github.com/szos/stumpwm-settings-gui) and [here](https://github.com/szos/stumpwm-settings). Heres a video demonstration: 

https://user-images.githubusercontent.com/39150985/111913716-b0f85600-8a6f-11eb-8947-682e7743e9d6.mp4

**Comment by dmb2** [comment]
@szos I want the gui in stumpwm proper. it looks amazing!

**Comment by szos** [comment]
@dmb2 Im glad you like it! I think having the gui in stumpwm proper would be a big step dependency-wise. It would mean adding defconfig, slim, and McClim as dependencies. McClim in particular is rather hefty. Would it be possible to add a way of `make`-ing stumpwm with the system, but allowing those who dont want it to avoid having it as part of the stumpwm binary? eg `$ make no-customize` or something similar? I know @fourier opined that the customization system seemed more complicated than it was worth. 

McClim has also given me issues in the past with things breaking unexpectedly, so I think having an option to build stumpwm without those dependencies would be useful. Ill take a look at the stumpwm makefile and see what I can discern.

**Comment by szos** [comment]
@dmb2 As requested Ive integrated this into stumpwm proper in my branch [cust](https://github.com/szos/stumpwm/tree/cust). Check it out and let me know if i should open a PR. Given that it adds McCLIM, SLIM, and Defconfig as dependencies I'm unsure if this should be added without a way of ignoring it and keeping the dependencies to only `clx`, `cl-ppcre`, and `alexandria`. Let me know what you think!

**Comment by sdilts** [comment]
I would suggest including as a module the main repo. That way it would ship
with the rest of the code, but could be off and unloaded by default.
